File: LC/[1326] Minimum Number of Taps to Open to Water a Garden.py
class Solution:
    def minTaps(self, n: int, ranges: List[int]) -> int:
        """
        1. recognize this is an interval question
        => min num of intervals to cover range -> sort by start point
        => max num of intervals of non-overlapping intervals 
        """
        N=n
        # convert to intervals
        intervals=[(ii-ranges[ii], ii+ranges[ii]) for ii in range(n+1)]
        # sort by starting points
        intervals.sort(key=lambda x:[x[0], -x[1]])
        # for each end_point, find max next end_point
        cnt=0 # count of intervals
        far=0 # 0 is the first end_point we identify in scope intervals and find next max end_point
        # ending criteria: end_point is not updating
        # endpoint is not exceeding end of range
        ii=0
        while ii < N:
            cnt +=1
            next_far=far
            while ii <N and intervals[ii][0]<=far:
                next_far=max(next_far, intervals[ii][1])
                ii+=1
            if next_far >= N: return cnt
            if next_far==far: return -1
            far=next_far
     

        return -1

File: LC/[1326] Minimum Number of Taps to Open to Water a Garden.cpp
class Solution {
    static bool cmp(array<int,2>&a, array<int,2>&b)
    {
        if (a[0]!=b[0])
            return a[0] < b[0];
        else
            return a[1] > b[1];
    }
public:
    int minTaps(int n, vector<int>& ranges) 
    {
        vector<array<int,2>>intervals(n+1);
        for (int i=0; i<=n; i++)
            intervals[i] = {i-ranges[i], i+ranges[i]};
        sort(intervals.begin(), intervals.end(), cmp);

        if (intervals[0][0]>0) return -1; //no interval covering the starting point

        int i = 0;
        int far = 0;
        int count = 0;
        while (i<intervals.size()) //traversing all sorted intervals
        {                        
            count+=1;
            int nextFar = far;
            //all in scope intervals that starts on or before far
            while (i<intervals.size() && intervals[i][0]<=far)
            {
                //find the interval that is with the new max far
                nextFar = max(nextFar, intervals[i][1]);
                i++;
            }            
            if (nextFar >= n) //ending point check
                return count;
            if (nextFar == far) //unable to push forward -> we are done
                return -1;
            far = nextFar;
        }

        return -1;
    }
};